src/app/screen.cpp

```diff
@@ -40,31 +47,65 @@ screen::screen(rmioc::screen& device, rfbClient* vnc_client)
     this->vnc_client->GotFrameBufferUpdate = screen::update_framebuf;
 }
 
+void screen::repaint(bool direct)
+{
+    /* If the update is direct, we don't clear the has_update flag
+          * Since direct updates only update black pixel, we still need to do a proper update every once in a whil
+          */
+    if(!direct) {
+       this->update_info.has_update = false;
+       this->update_info.last_repaint_time = chrono::steady_clock::now();
+    }
+    log::print("Screen update")
+        << this->update_info.w << 'x' << this->update_info.h << '+'
+        << this->update_info.x << '+' << this->update_info.y << '\n';
+
+    this->device.update(
+           this->update_info.x, this->update_info.y,
+           this->update_info.w, this->update_info.h,
+           direct
+     );
+}
+int screen::get_xres()
+{
+    return this->device.get_xres();
+}
+int screen::get_yres()
+{
+    return this->device.get_yres();
+}
+ 
+void screen::set_repainting_mode(repainting_mode mode)
+{
+    this->repaint_mode = mode;
+}
+
+   
 auto screen::event_loop() -> event_loop_status
 {
     if (this->update_info.has_update)
     {
+        auto now = chrono::steady_clock::now();
         int remaining_wait_time =
             chrono::duration_cast<chrono::milliseconds>(
                 this->update_info.last_update_time + update_delay
-                - chrono::steady_clock::now()
+                - now
             ).count();
 
-        if (remaining_wait_time <= 0)
+        int must_repaint = 
+            chrono::duration_cast<chrono::milliseconds>(
+                this->update_info.last_repaint_time + update_max_delay
+                - now
+            ).count();
+        if (remaining_wait_time <= 0 || must_repaint <= 0)
         {
-            this->update_info.has_update = false;
-
-            log::print("Screen update")
-                << this->update_info.w << 'x' << this->update_info.h << '+'
-                << this->update_info.x << '+' << this->update_info.y << '\n';
-
-            this->device.update(
-                this->update_info.x, this->update_info.y,
-                this->update_info.w, this->update_info.h
-            );
+            this->repaint();
         }
         else
         {
+            if (this->repaint_mode == app::repainting_mode::fast)
+                this->repaint(true);
+
             // Wait until the next update is due
             return {/* quit = */ false, /* timeout = */ remaining_wait_time};
```

Comment:
Shouldn’t this be changed to

```suggestion
            return {/* quit = */ false, /* timeout = */ std::min(remaining_wait_time, must_repaint)};
```

So that the next update is scheduled as soon as one of the two conditions are met?

Reply:
True, I did not spot out this !

Reply:
Thanks!